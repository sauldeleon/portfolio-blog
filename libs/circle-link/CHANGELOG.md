# Changelog Circle Link component

## [0.2.1](https://github.com/sauldeleon/portfolio-blog/compare/circle-link-0.2.0...circle-link-0.2.1) (2023-09-14)

# [0.2.0](https://github.com/sauldeleon/portfolio-blog/compare/circle-link-0.1.2...circle-link-0.2.0) (2023-09-10)

### Bug Fixes

- on icon focus bounce portals ([486a64a](https://github.com/sauldeleon/portfolio-blog/commit/486a64a0a6085ba6e987dee4cc20702c208cdf1b))

### Features

- add first version of about modal ([c876e99](https://github.com/sauldeleon/portfolio-blog/commit/c876e99058a6167230f85c006d17624fe7a7c456))

## [0.1.2](https://github.com/sauldeleon/portfolio-blog/compare/circle-link-0.1.1...circle-link-0.1.2) (2023-09-08)

### Bug Fixes

- improve a11y of skill icons ([092b579](https://github.com/sauldeleon/portfolio-blog/commit/092b579c9b2adb616662a7018d1023fe01668662))

## [0.1.1](https://github.com/sauldeleon/portfolio-blog/compare/circle-link-0.1.0...circle-link-0.1.1) (2023-09-05)

# 0.1.0 (2023-09-02)

### Bug Fixes

- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- add button component ([b8f03d5](https://github.com/sauldeleon/portfolio-blog/commit/b8f03d549e75ca2055a076c7b06416b91bbf00f4))
- add button with hovering effect component ([4283ceb](https://github.com/sauldeleon/portfolio-blog/commit/4283ceb3329722ce4f6c856da48a9e4a4f65c36b))
- add footer ([217e4f7](https://github.com/sauldeleon/portfolio-blog/commit/217e4f726a6d491cf3d0c8d088c99646960b98f3))
- add keyframes helpers ([5f35d60](https://github.com/sauldeleon/portfolio-blog/commit/5f35d60841de4e459298d705ea25b16c63dba4a6))
- add link component ([d39b47d](https://github.com/sauldeleon/portfolio-blog/commit/d39b47dba6c1cc14982aa66d8e0375b7357ff1cc))
- add portal game ([47c1ea2](https://github.com/sauldeleon/portfolio-blog/commit/47c1ea27030f88479a1f7eab1bc15842c960725c))
- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))
- improved a11y and styles ([873f656](https://github.com/sauldeleon/portfolio-blog/commit/873f656587f50df8e29ad526fdc68fea0b2c9784))
