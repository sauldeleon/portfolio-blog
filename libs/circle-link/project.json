{
  "name": "circle-link",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/circle-link/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "version": {
      "executor": "@jscutlery/semver:version",
      "options": {
        "changelogHeader": "# Changelog Circle Link component \n",
        "baseBranch": "main",
        "commitMessageFormat": "chore({projectName}): release: v{version}",
        "push": false
      }
    },
    "lint": {
      "executor": "@nx/eslint:lint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/circle-link/**/*.{ts,tsx,js,jsx}"]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/circle-link/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  }
}
