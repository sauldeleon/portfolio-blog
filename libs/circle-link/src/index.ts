export * from './lib/circle-link'
