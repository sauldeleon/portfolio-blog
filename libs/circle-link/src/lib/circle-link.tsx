import React from 'react'

import {
  StyledBody,
  StyledCircleButtonIcon,
  StyledCircleLink,
  StyledIconContent,
  StyledIconWrapper,
} from './circle-link.styles'

export interface CircleLinkProps
  extends React.AnchorHTMLAttributes<HTMLAnchorElement> {
  label?: React.ReactNode
  iconContent?: React.ReactNode
  iconSize?: number
}

export function CircleLink({
  label,
  iconContent,
  iconSize = 42,
  ...rest
}: CircleLinkProps) {
  return (
    <StyledCircleLink $size={iconSize} {...rest}>
      <StyledIconWrapper $size={iconSize}>
        <StyledCircleButtonIcon width={iconSize} height={iconSize} />
        {iconContent && <StyledIconContent>{iconContent}</StyledIconContent>}
      </StyledIconWrapper>
      {label && <StyledBody>{label}</StyledBody>}
    </StyledCircleLink>
  )
}
