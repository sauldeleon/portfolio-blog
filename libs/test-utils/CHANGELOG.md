# Changelog Test utils

## [0.2.3](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.2...test-utils-0.2.3) (2024-01-08)

## [0.2.2](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.1...test-utils-0.2.2) (2023-09-26)

## [0.2.1](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.0...test-utils-0.2.1) (2023-09-14)

# [0.2.0](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.1.0...test-utils-0.2.0) (2023-09-05)

### Bug Fixes

- reduced duplicated lines ([f51168a](https://github.com/sauldeleon/portfolio-blog/commit/f51168aefe025a5a65ffdb2dc22f7a16b863406e))
- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))

# 0.1.0 (2023-09-02)

### Bug Fixes

- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- add contact initial page ([fea862b](https://github.com/sauldeleon/portfolio-blog/commit/fea862b3507268013149050ae3f05fa517e7b3d9))
- add homepage ([60523aa](https://github.com/sauldeleon/portfolio-blog/commit/60523aa31ec088d318745ef0c757b4cbe656f0d4))
- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))
- improved a11y and styles ([873f656](https://github.com/sauldeleon/portfolio-blog/commit/873f656587f50df8e29ad526fdc68fea0b2c9784))
- toggle language button ([afd97e9](https://github.com/sauldeleon/portfolio-blog/commit/afd97e95beac0f70ed643c753e71db2ca014054b))
