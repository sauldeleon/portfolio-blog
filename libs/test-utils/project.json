{
  "name": "test-utils",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/test-utils/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "version": {
      "executor": "@jscutlery/semver:version",
      "options": {
        "changelogHeader": "# Changelog Test utils \n",
        "baseBranch": "main",
        "commitMessageFormat": "chore(${projectName}): release: v${version}",
        "push": false
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/test-utils/**/*.{ts,tsx,js,jsx}"]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/test-utils/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  }
}
