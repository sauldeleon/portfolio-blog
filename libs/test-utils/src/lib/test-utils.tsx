import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { Matcher } from '@testing-library/dom/types/matches'
import {
  RenderOptions,
  RenderResult,
  queryAllByAttribute as _queryAllByAttribute,
  queryByAttribute as _queryByAttribute,
  act,
  render,
  screen,
} from '@testing-library/react'
import { RouterContext } from 'next/dist/shared/lib/router-context'
import Router, { NextRouter } from 'next/router'
import { ReactNode, useEffect, useState } from 'react'
import { ThemeProvider } from 'styled-components'

import { LanguageContextProvider } from '@sdlgr/i18n-config'
import { mainTheme } from '@sdlgr/main-theme'

export type MatcherCustomFnParams = {
  content: string
  element: Element | null
}

/* istanbul ignore next */
export const renderWithTheme = (children: ReactNode): RenderResult => {
  return render(<ThemeProvider theme={mainTheme}>{children}</ThemeProvider>)
}

/* istanbul ignore next */
export function RenderProviders({ children }: { children: ReactNode }) {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

  return (
    <QueryClientProvider client={queryClient}>
      <LanguageContextProvider value={{ language: 'en' }}>
        <ThemeProvider theme={mainTheme}>{children}</ThemeProvider>
      </LanguageContextProvider>
    </QueryClientProvider>
  )
}

/* istanbul ignore next */
export const renderApp = (
  children: ReactNode,
  options?: Omit<RenderOptions, 'queries'>
) => {
  return render(<RenderProviders>{children}</RenderProviders>, options)
}

/**
 * Generic function to get element by attribute and value
 * @example queryByAttribute('id')(baseElement, 'element-id')
 * @param {string} attr
 * @return {any}
 */
export function queryByAttribute(attr: string) {
  return _queryByAttribute.bind(null, attr)
}

/**
 * Generic function to get all elements by attribute and value
 * @example queryByAttribute('class')(baseElement, 'element-class')
 * @param {string} attr
 * @returns {any}
 */
export function queryAllByAttribute(attr: string) {
  return _queryAllByAttribute.bind(null, attr)
}

/**
 * Generic function to get element by attribute and value
 * @example getByAttribute('id')(baseElement, 'element-id')
 * @param {string} attr
 * @return {any}
 */
export function getByAttribute(attr: string) {
  return _queryByAttribute.bind(null, attr)
}

/**
 * Generic function to get all elements by attribute and value
 * @example getByAttribute('class')(baseElement, 'element-class')
 * @param {string} attr
 * @returns {any}
 */
export function getAllByAttribute(attr: string) {
  return _queryAllByAttribute.bind(null, attr)
}

/**
 * Shortcut version for queryByAttribute, but specific for id
 * @param {HTMLElement} container
 * @param {string} id
 * @return {any}
 */
export function getById(container: HTMLElement, id: string) {
  return queryByAttribute('id')(container, id) as Element
}

/**
 * Shortcut version for queryByAttribute, but specific for class name
 * @param {HTMLElement} container
 * @param {string} className
 * @return {any}
 */
export function getByClassName(container: HTMLElement, className: string) {
  return queryByAttribute('class')(container, className) as Element
}

function matcherCustomFn(
  tag: string,
  matcher: Matcher,
  { element, content }: MatcherCustomFnParams
) {
  const isTag = element?.tagName.toUpperCase() === tag.toUpperCase()
  let matching = false
  if (isTag) {
    matching = content === matcher
    /* istanbul ignore else*/
    if (typeof matcher === 'function') {
      matching = matcher(content, element)
    } else if (matcher instanceof RegExp) {
      matching = matcher.test(content)
    }
  }
  return matching && isTag
}

/**
 * Get 1 element by tag and text
 * @param {string} tag
 * @param {Matcher} matcher
 * @return {any}
 */
export function getByTagAndText(tag: string, matcher: Matcher) {
  return screen.getByText((content, element) =>
    matcherCustomFn(tag, matcher, { content, element })
  )
}

/**
 * Get multiple elements by tag and text
 * @param {string} tag
 * @param {Matcher} matcher
 * @return {any}
 */
export function getAllByTagAndText(tag: string, matcher: Matcher) {
  return screen.getAllByText((content, element) =>
    matcherCustomFn(tag, matcher, { content, element })
  )
}

/**
 * Mock Math.random() to return a specific value
 * @param {number} value
 */
export function mockRandom(value = 0.1234) {
  const mockMath = Object.create(global.Math)
  mockMath.random = () => value
  global.Math = mockMath
}

/**
 * Hook to use in test to ensure "last" render is called
 */
export function useWithRefreshInTest() {
  const [withRefreshInTest, setWithRefreshInTest] = useState(false)

  useEffect(() => {
    // Change state to "rerender" the component
    setWithRefreshInTest(true)
  }, [])

  return withRefreshInTest
}

export { render } from '@testing-library/react'

/* istanbul ignore next */
export const awaitAllAsync = async () =>
  act(() => new Promise((resolve) => setTimeout(resolve, 0)))

/* istanbul ignore next */
export function renderWithNextRouter(
  children: ReactNode,
  options?: Partial<NextRouter>
) {
  return renderWithTheme(
    <RouterContext.Provider value={{ ...Router, ...options }}>
      {children}
    </RouterContext.Provider>
  )
}
