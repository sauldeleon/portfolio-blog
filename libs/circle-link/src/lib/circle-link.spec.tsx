import { screen } from '@testing-library/react'

import { renderWithTheme } from '@sdlgr/test-utils'

import { CircleLink } from './circle-link'

describe('CircleLink', () => {
  it('should render successfully', () => {
    const { baseElement } = renderWithTheme(<CircleLink href="" />)
    expect(baseElement).toBeTruthy()
  })

  it('should render with a label', () => {
    renderWithTheme(<CircleLink label="test" href="" />)
    expect(screen.getByText('test')).toBeInTheDocument()
  })

  it('should render with a label inside the icon', () => {
    renderWithTheme(<CircleLink label="test" iconContent="wadus" href="" />)
    expect(screen.getByText('wadus')).toBeInTheDocument()
  })

  it('should render with a different size', () => {
    const { baseElement } = renderWithTheme(
      <CircleLink label="test" iconContent="wadus" iconSize={50} href="" />,
    )
    expect(baseElement).toMatchSnapshot()
  })

  it('should render a link with given props', () => {
    renderWithTheme(
      <CircleLink
        label="test"
        iconContent="wadus"
        iconSize={50}
        href="/path"
      />,
    )

    expect(screen.getByRole('link')).toHaveAttribute('href', '/path')
  })
})
