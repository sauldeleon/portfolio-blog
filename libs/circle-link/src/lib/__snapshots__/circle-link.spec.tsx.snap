// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircleLink should render with a different size 1`] = `
.c0 {
  position: relative;
  position: relative;
}

.c0::after {
  content: '';
  position: absolute;
  width: 100%;
  height: 2px;
  bottom: 0;
  left: 0;
  background-color: currentColor;
}

.c0::after {
  transform: scaleX(0);
  transform-origin: bottom right;
  transition: transform 0.25s ease-out;
}

.c0:hover::after {
  transform: scaleX(1);
  transform-origin: bottom left;
}

.c0:hover,
.c0:active,
.c0:focus {
  text-decoration: none;
}

.c6 {
  font-family: 'Roboto mono',monospace;
  font-weight: 400;
  color: #FBFBFB;
  font-size: 0.9rem;
  line-height: normal;
  white-space: pre-line;
}

.c6 ul,
.c6 ol {
  margin-left: 40px;
  margin-bottom: 10px;
}

.c6 ul {
  list-style: outside disc;
}

.c6 ol {
  list-style: decimal;
}

.c6 li {
  padding-left: 10px;
}

.c7 {
  margin-left: 8px;
  position: relative;
}

.c7::after {
  content: '';
  position: absolute;
  width: 100%;
  height: 2px;
  bottom: 0;
  left: 0;
  background-color: currentColor;
}

.c7::after {
  transform: scaleX(0);
  transform-origin: bottom right;
  transition: transform 0.5s ease-out;
}

.c1 {
  outline: none;
  cursor: pointer;
  display: flex;
  align-items: center;
  max-height: 50px;
}

.c1::after {
  content: none;
}

.c1:hover circle.loading {
  stroke: white;
  stroke-width: 6px;
  stroke-dasharray: 110;
  stroke-dashoffset: 82;
  animation: kVEmCK 0.5s steps(4,jump-none) forwards;
  transform: rotate(-90deg);
  transform-origin: center;
}

.c1:hover .c5::after {
  transform: scaleX(1);
  transform-origin: bottom left;
}

.c2 {
  position: relative;
  width: 50px;
  height: 50px;
}

.c3 {
  position: absolute;
  transform: translate(-50%,-50%);
  top: 50%;
  left: 50%;
}

.c4 {
  position: absolute;
  transform: translate(-50%,-50%);
  top: 50%;
  left: 50%;
  display: flex;
}

<body>
  <div>
    <a
      class="c0 c1"
      href=""
    >
      <div
        class="c2"
      >
        <svg
          class="c3"
          height="50"
          width="50"
        >
          circleButton.svg
        </svg>
        <div
          class="c4"
        >
          wadus
        </div>
      </div>
      <p
        class="c5 c6 c7"
      >
        test
      </p>
    </a>
  </div>
</body>
`;
